### PR TITLE
SAM mlflow converter/export

### PR DESCRIPTION
**Pull request by shubhamiit**
Adding mlflow export/conversion logic of SAM to model management repo.

Working endpoint - https://ml.azure.com/endpoints/realtime/sam-huge-2?wsid=/subscriptions/dbd697c3-ef40-488f-83e6-5ad4dfb78f9b/resourceGroups/grajguru-rg/providers/Microsoft.MachineLearningServices/workspaces/grajguru-ws&tid=72f988bf-86f1-41af-91ab-2d7cd011db47

pipeline run - https://ml.azure.com/experiments/id/310732e9-14db-4e9e-a1d7-d19233140c42/runs/f336a278-12d6-4e5c-9b19-b104b141e420?wsid=/subscriptions/dbd697c3-ef40-488f-83e6-5ad4dfb78f9b/resourceGroups/grajguru-rg/providers/Microsoft.MachineLearningServices/workspaces/grajguru-ws&tid=72f988bf-86f1-41af-91ab-2d7cd011db47#

image

### CODE REVIEW

File: assets/training/model_management/components/convert_model_to_mlflow/spec.yaml [hunk 1]
```diff
@@ -66,6 +66,7 @@ inputs:
       - text-to-image-inpainting
       - image-text-to-text
       - zero-shot-image-classification
+      - segment-anything
```
Comment (rjaincc) [inline assets/training/model_management/components/convert_model_to_mlflow/spec.yaml:69]: task type --> mask-generation

File: assets/training/model_management/src/azureml/model/mgmt/processors/common/vision_utils.py [hunk 2]
```diff
@@ -187,11 +192,48 @@ def get_current_device() -> torch.device:
...
+
+def bool_array_to_pil_image(bool_array):
+    """Convert boolean array to PIL Image."""
+    # Convert boolean array to uint8
+    uint8_array = bool_array.astype(np.uint8) * 255
+
+    # Create a PIL Image
+    pil_image = Image.fromarray(uint8_array)
+
+    return pil_image
```
Comment (rjaincc) [inline assets/training/model_management/src/azureml/model/mgmt/processors/common/vision_utils.py:239]: nit: empty line at the end of the file

File: assets/training/model_management/src/azureml/model/mgmt/processors/pyfunc/segment_anything/segment_anything_mlflow_wrapper.py [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+        context: mlflow.pyfunc.PythonModelContext,
+        input_data: pd.DataFrame,
+    ) -> pd.DataFrame:
+        """
+        Perform inference on the input data.
+
+        :param context: MLflow context containing artifacts that the model can use for inference
+        :type context: mlflow.pyfunc.PythonModelContext
+        :param input_data:
+            Pandas DataFrame with a first column name ["image"] which is in base64 String format or an URL, and
```
Comment (rjaincc) [inline assets/training/model_management/src/azureml/model/mgmt/processors/pyfunc/segment_anything/segment_anything_mlflow_wrapper.py:68]: nit: the desc for input_data has weird \n(s).

File: assets/training/model_management/components/convert_model_to_mlflow/spec.yaml (hunk 1 above)
Comment (shubhamiit) [reply assets/training/model_management/components/convert_model_to_mlflow/spec.yaml:69]: updated

File: assets/training/model_management/src/azureml/model/mgmt/processors/pyfunc/segment_anything/segment_anything_mlflow_wrapper.py (hunk 3 above)
Comment (shubhamiit) [reply assets/training/model_management/src/azureml/model/mgmt/processors/pyfunc/segment_anything/segment_anything_mlflow_wrapper.py:68]: Didn't get it, can you open it via view file option and point out the area?

Comment (rjaincc) [reply assets/training/model_management/src/azureml/model/mgmt/processors/pyfunc/segment_anything/segment_anything_mlflow_wrapper.py:69]: image

File: assets/training/model_management/src/azureml/model/mgmt/processors/common/vision_utils.py [hunk 4]
```diff
@@ -187,11 +192,48 @@ def get_current_device() -> torch.device:
...
+            if isinstance(item, list):
+                to_float_recursive(item)
+            else:
+                lst[i] = float(item)
+
+    to_float_recursive(nested_list)
+    return nested_list
+
+
+def bool_array_to_pil_image(bool_array):
```
Comment (rjaincc) [inline assets/training/model_management/src/azureml/model/mgmt/processors/common/vision_utils.py:231]: doc string missing

File: assets/training/model_management/src/azureml/model/mgmt/processors/pyfunc/segment_anything/segment_anything_mlflow_wrapper.py [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+                                bool_array_to_pil_image(masks_numpy[i][j]), format=DatatypeLiterals.IMAGE_FORMAT
+                            ),
+                            MLflowSchemaLiterals.RESPONSE_DF_IOU_SCORE: scores_list[i][j],
+                        }
+                    )
+                pred[MLflowSchemaLiterals.RESPONSE_DF_PREDICTIONS].append(per_pred)
+            response.append(pred)
+
+        df_responses = pd.DataFrame({MLflowSchemaLiterals.OUTPUT_COLUMN_RESPONSE: response})
+        return df_responses
```
Comment (rjaincc) [inline assets/training/model_management/src/azureml/model/mgmt/processors/pyfunc/segment_anything/segment_anything_mlflow_wrapper.py:177]: nit: empty line at the end of the file.

File: assets/training/model_management/src/azureml/model/mgmt/processors/pyfunc/segment_anything/segment_anything_mlflow_wrapper.py (hunk 3 above)
Comment (shubhamiit) [reply assets/training/model_management/src/azureml/model/mgmt/processors/pyfunc/segment_anything/segment_anything_mlflow_wrapper.py:69]: Copied from official docstring: https://github.com/huggingface/transformers/blob/574a5384557b1aaf98ddb13ea9eb0a0ee8ff2cb2/src/transformers/models/sam/modeling_sam.py#L1132

File: assets/training/model_management/src/azureml/model/mgmt/processors/pyfunc/convertors.py [hunk 6]
```diff
@@ -563,3 +565,87 @@ def _prepare_artifacts_dict(self) -> Dict:
...
+        :return: MLflow model signature.
+        :rtype: mlflow.models.signature.ModelSignature
+        """
+        input_schema = Schema(
+            [
+                ColSpec(
+                    SegmentAnythingMLFlowSchemaLiterals.INPUT_COLUMN_IMAGE_DATA_TYPE,
+                    SegmentAnythingMLFlowSchemaLiterals.INPUT_COLUMN_IMAGE,
+                ),
+                ColSpec(
```
Comment (vivek-dani) [inline assets/training/model_management/src/azureml/model/mgmt/processors/pyfunc/convertors.py:594]: No signature would still be more user friendly than specifying signature, where we'd have to give list as string

Comment (shubhamiit) [reply assets/training/model_management/src/azureml/model/mgmt/processors/pyfunc/convertors.py:594]: @vivek-dani Yes, testing it without no-signature.

File: assets/training/model_management/src/azureml/model/mgmt/processors/pyfunc/segment_anything/config.py [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+    MASK_GENERATION = "mask-generation"
+
+
+class MLflowLiterals:
+    """MLflow export related literals."""
+
+    MODEL_DIR = "model_dir"
+
+
+class MLflowSchemaLiterals:
```
Comment (MadhuM02) [inline assets/training/model_management/src/azureml/model/mgmt/processors/pyfunc/segment_anything/config.py:28]: can we keep all MLFLOW related constants at a common place and use them??

File: assets/training/model_management/src/azureml/model/mgmt/processors/pyfunc/segment_anything/segment_anything_mlflow_wrapper.py [hunk 8]
```diff
@@ -0,0 +1,178 @@
...
+            fifth column name ["multimask_output"] -
+            boolean value to indicate if the model should return a single mask or multiple masks per input point.
+            If False, the model will return a single mask per input point, if True, the model will return multiple
+            masks per input point. Default is True.
+        :type input_data: pd.DataFrame
+        :return: Pandas dataframe with generated masks and IOU score corresponding to each inputs
+            [input_points, input_boxes, input_labels] converted to base64 string.
+        :rtype: pd.DataFrame
+        """
+        # Do inference one input at a time.
```
Comment (MadhuM02) [inline assets/training/model_management/src/azureml/model/mgmt/processors/pyfunc/segment_anything/segment_anything_mlflow_wrapper.py:118]: why only one image at a time??

File: assets/training/model_management/src/azureml/model/mgmt/processors/pyfunc/convertors.py (hunk 6 above)
Comment (shubhamiit) [reply assets/training/model_management/src/azureml/model/mgmt/processors/pyfunc/convertors.py:672]: Summary of testing with No-Signature:

Online Endpoint- With no signature, Json was read directly as pandas dataframe and endpoint was able to pass list of list of integers to the predict function as dataframe.

Batch Endpoint - In batch, with no signature case, csv file was read directly as pandas dataframe which expect string conversion of list of list of integers and predict functions gets a string in the case which is different from online endpoint.

Conclusion/Action- Decided to keep the backend logic simple and common for both batch and online endpoint, will take string as input for both and convert them to list of list of integers as of now.

File: assets/training/model_management/src/azureml/model/mgmt/processors/pyfunc/segment_anything/segment_anything_mlflow_wrapper.py (hunk 5 above)
Comment (shubhamiit) [reply assets/training/model_management/src/azureml/model/mgmt/processors/pyfunc/segment_anything/segment_anything_mlflow_wrapper.py:178]: Needed for flake8

File: assets/training/model_management/src/azureml/model/mgmt/processors/common/vision_utils.py [hunk 9]
```diff
@@ -187,11 +192,57 @@ def get_current_device() -> torch.device:
...
+    # Check if string matches the expected format using a regex pattern.
+    # This pattern ensures that the string only contains brackets, numbers, dots, and commas.
+    if not re.match(r"^\[((\[|\]|\s|\d|\,|\.|-)*?)\]$", s):
+        raise ValueError("Invalid input format, please use a nested list of ints or floats.")
+
+    # Convert string to nested list using literal_eval
+    nested_list = literal_eval(s)
+
+    # Helper function to convert nested lists to floats
+    def to_float_recursive(lst):
```
Comment (rjaincc) [inline assets/training/model_management/src/azureml/model/mgmt/processors/common/vision_utils.py:223]: doc string?

**Review by rjaincc — DISMISSED** [review]

**Review by MadhuM02 — APPROVED** [review]

**Review by rjaincc — APPROVED** [review]

**Review by MadhuM02 — APPROVED** [review]

**Pull request merged**